Restructure SEC company information tables

katie-lamb opened the pull request:
<!--
Resources:
* contributing guidelines: https://catalystcoop-pudl.readthedocs.io/en/nightly/CONTRIBUTING.html
* code of conduct: https://catalystcoop-pudl.readthedocs.io/en/nightly/code_of_conduct.html
-->

# Overview

Closes #4078 . The first of two (or more) SEC table restructuring PRs - this handles the quarterly filings table and company information tables. It doesn't include the ownership tables.

## What problem does this address?

Makes changes to the SEC table structures to be more well normalized and usable.

## What did you change?

**core_sec10k__quarterly_filings**
This table is largely the same. Only minor changes were made.

- [x] Ensure that the CIK of each filing matches the CIK from the filename string - make a test for this?
- [x] Update `filing_date` column description to indicate that it's daily frequency
- [x] Indicate that `report_date` is the quarter that the `filing_date` pertains to
- [x] Enforce format of `exhibit_21_version` with a regex - enforce this with a constraint in the field metadata?

#### Company Information Tables
**raw_sec10k__quarterly_company_information**

- [x] Rename `core_sec10k__quarterly_company_information` to be a raw table

**core_sec10k__quarterly_company_information**

- [x] Pivot the raw table so that field values are columns
- [x] Make `report_date` and `central_index_key` the primary key
- [x] Strip leading "[" from field values
- [x] Add field metadata for new columns
- [x] Investigate how many CIKs from the quarterly filings table don't have corresponding harvested company information in this table.
  * I see 257/39183 filer CIKs that aren't harvested into this company information table coming from 360 filings. It looks like ~200 of these filings are pre-2000 report dates. See [this comment](https://github.com/catalyst-cooperative/pudl/issues/4078#issuecomment-2683706842).
- [x] Update row counts in the ETL fast and full row counts CSVs

Questions:
* `central_index_key` and `report_date` are not the natural primary key because there are slight differences in headers harvested from different filings. I prioritized headers from filings where the filer is the same as the record we're harvesting but we do lose some records by forcing the CIK + report date primary key. Any reason to leave filename as the primary key?

**out_sec10k__quarterly_company_information**

- [x] Merge `utility_id_eia` and `utility_name_eia` onto the core company information table.

**core_sec10k__changelog_company_name**

- [x] Pull `central_index_key`, `report_date`, `name_change_date`, `former_conformed_name`, `current_name` out of raw company information table and see if `central_index_key`, `report_date`, `name_change_date`is a natural primary key
- [ ] make FK relationship back to core table

# Documentation

Make sure to update relevant aspects of the documentation.

```[tasklist]
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.
- [ ] Update relevant Data Source jinja templates (see `docs/data_sources/templates`).
- [x] Update relevant table or source description metadata (see `src/metadata`).
- [x] Review and update any other aspects of the documentation that might be affected by this PR.
```

# Testing

## How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [x] If updating analyses or data processing functions: make sure to update or write data validation tests (e.g. `test_minmax_rows()`).
- [ ] Run `make pytest-coverage` locally to ensure that the merge queue will accept your PR.
- [x] Review the PR yourself and call out any questions or issues you have.
- [ ] For minor ETL changes or data additions, once `make pytest-coverage` passes, make sure you have a fresh full PUDL DB downloaded locally, materialize new/changed assets and all their downstream assets and [run relevant data validation tests](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `pytest` and `--live-dbs`.
- [ ] For bigger ETL or data changes run the full ETL locally and then [run the data validations](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `make pytest-validate`.
- [ ] Alternatively, run the `build-deploy-pudl` GitHub Action manually.
```

katie-lamb commented on dbt/models/sec10k/core_sec10k__quarterly_company_information/schema.yml:
```diff
@@ -8,10 +8,24 @@ sources:
               table_name: core_sec10k__quarterly_company_information
               partition_column: report_date
         columns:
```
I think I'm supposed to update these `dbt` schema files, but I'm unfamiliar with all the dbt stuff so not sure I did this right.

katie-lamb replied:
Oh sorry, my initial comment was phrased poorly. I meant - I updated these schema files, but I wasn't sure if I was actually supposed to. 

Seems like the rest of the comment is directed at @zschira and is more about formatting and diffs. Could be moved to a separate issue?

katie-lamb commented on dbt/seeds/etl_fast_row_counts.csv:
```diff
@@ -686,36 +686,36 @@ core_pudl__codes_datasources,,11
-core_sec10k__quarterly_company_information,2021,195784
-core_sec10k__quarterly_company_information,2022,202021
-core_sec10k__quarterly_company_information,2023,193870
+core_sec10k__quarterly_company_information,1994,2286
```
The relative row counts of years seems to be consistent with the previous (much larger) version of this table. It makes sense to me that this table is now much smaller.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -39,13 +43,121 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+    raw_df["company_information_fact_name"] = raw_df[
+        "company_information_fact_name"
+    ].str.strip("]")
+    raw_df["company_information_block"] = pd.Categorical(
+        raw_df["company_information_block"],
+        [
+            "business_address",
+            "mail_address",
+            "company_data",
+            "filing_values",
+            "former_company",
+        ],
+    )
+    df = raw_df.sort_values("company_information_block").pivot_table(
+        values="company_information_fact_value",
+        index=["filename_sec10k", "report_date"],
+        columns="company_information_fact_name",
+        aggfunc="first",
+    )
```
I started by trying to just pivot the raw table to see if `report_date` and `filename_sec10k` was a natural primary key. It wasn't, because there are instances where similar information is reported in a different `company_information_block` for the same filename and report date. From some spot checking, it seemed like the differences between the information reported in these blocks was usually pretty small (e.g. the `film_number` was different), so I decided to only keep one of the blocks.

I chose a somewhat arbitrary hierarchy of blocks - business address, mail address, company data, filing values, former company - based on the relative counts of these block types in the raw data. Then, I sorted the data based on this hierarchy and pivoted while keeping only the "first" instance of each unique `report_date` and `filename_sec10k` index value.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -39,13 +43,121 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+    core_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Changes in SEC company names and the date of change as reported in 10k filings."""
+    changelog_df = core_df[
+        [
+            "central_index_key",
+            "report_date",
+            "company_name",
+            "name_change_date",
+            "company_name_former",
+        ]
+    ]
```
Maybe I misunderstood this table, but basically it's just pulling the recorded name changes and dates out of the core (pivoted) table? Or I guess this could be made more robust by actually tracking when the name changes for each CIK even if the company doesn't report a name change.

zaneselvans replied:
I was imagining this just representing the name changes as they're explicitly reported -- not checking against all the report dates and company names that are observed to try and detect changes.  That's a good idea! But might not be worth the effort, and in any case we would first need a straightforward extraction of the reported name changes to check against the empirical name changes.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+This table is pulled from ``core_sec10k__quarterly_company_information`` and contains
+data extracted from SEC 10k filings."""
+        ),
+        "schema": {
+            "fields": [
+                "central_index_key",
+                "report_date",
+                "company_name",
+                "name_change_date",
+                "company_name_former",
+            ],
+            "primary_key": ["central_index_key", "report_date"],
+        },
```
Would adding a foreign key relationship to the core company information table just be on the `central_index_key` and `report_date`?

zaneselvans replied:
Typically a FK constraint points to the PK columns in another table, so if those were the PKs in the company info table, that's probably be what you want.  But I've left comments both on the natural PK of the company info table, and on what a better normalized version of this table might look like.

katie-lamb commented:
@zaneselvans The two unchecked boxes in the top level comment (enforce `exhibit_21_version` with a regex and add a FK relationship for the changelog table) haven't been addressed yet. Also, I'm not sure why my alembic migrations are working locally (I had to wipe out all previous migrations to get it working) but not in the GitHub CI.

zaneselvans commented:
On the alembic error it seems like the current migration is trying to drop a table that doesn't exist in the prior version of the database: `_alembic_tmp_core_sec10k__quarterly_company_inform`.

My guess would be that maybe you didn't downgrade your local database schema to the base version on `main` before creating the migration, so the diff that the migration encodes is actually between some intermediary database schema and the final schema you've got now?

zaneselvans commented:
The `pudl.sqlite` database which you were attempting to create a migration from did not seem to correspond to any version of the database which alembic was aware of. To get a clean migration, you need to start with a DB that alembic knows about. Otherwise the chain of migrations will be discontinuous. To make this work I moved my `pudl.sqlite` aside, removed the bad migration, ran `alembic upgrade head` to create an empty DB with the right schema, and then created a new migration from that guaranteed to be valid DB to the new DB schema represented by the metadata in PUDL.

Then I was able to run the ETL for the SEC 10-K tables, but the schema check on `core_sec10k__quarterly_company_information` failed, so it seems like something needs to be debugged.  Unfortunately, the error message was not very helpful:

```

Stack Trace:
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/dagster/_core/execution/plan/execute_plan.py", line 245, in dagster_event_sequence_for_step
    yield from check.generator(step_events)
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/dagster/_core/execution/plan/execute_step.py", line 501, in core_dagster_event_sequence_for_step
    for user_event in _step_output_error_checked_user_event_sequence(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/dagster/_core/execution/plan/execute_step.py", line 184, in _step_output_error_checked_user_event_sequence
    for user_event in user_event_sequence:
                      ^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/dagster/_core/execution/plan/execute_step.py", line 89, in _process_asset_results_to_events
    yield from _process_user_event(step_context, user_event)
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/dagster/_core/execution/plan/execute_step.py", line 133, in _process_user_event
    raise DagsterAssetCheckFailedError(
```

@jdangerx is there more informative error information somewhere I'm not seeing / forgetting about?

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -27,8 +31,8 @@ def _year_quarter_to_date(year_quarter: pd.Series) -> pd.Series:
     io_manager_key="pudl_io_manager",
     group_name="pudl_models",
```
So you mean just update the group name to `sec10k` and make sure that the table names correspond to the raw/core/out naming convention?

zaneselvans replied:
That would help them be easier to find when navigating asset groups in the Dagster UI for sure. But I think we could also just move them under the `extract`, `transform`, and `load` or `output` directories with modules named `sec10k.py` so that they follow the same naming & layout as all the other data sources.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -27,8 +31,8 @@ def _year_quarter_to_date(year_quarter: pd.Series) -> pd.Series:
 )
-def core_sec10k__quarterly_company_information() -> pd.DataFrame:
-    """Basic company information extracted from SEC10k filings."""
+def raw_sec10k__quarterly_company_information() -> pd.DataFrame:
+    """Raw company information harvested from headers of SEC10k filings."""
```
changed to not write raw tables to the DB.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -39,13 +43,121 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+@asset(
+    io_manager_key="pudl_io_manager",
+    group_name="pudl_models",
+    ins={"raw_df": AssetIn("raw_sec10k__quarterly_company_information")},
```
Removed `ins` in all places and used the asset name directly.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "organization_name",
+                "sec_act",
+                "sec_file_number",
+                "industry_id_sic",
```
There was an existing field called `industry_description_sic` (I think probably put in by the initial SEC PR) that could be changed to `industry_name_sic`.

katie-lamb commented on src/pudl/metadata/fields.py:
```diff
@@ -1724,7 +1724,11 @@
-        "description": "Date filing was submitted.",
+        "description": "Date filing was submitted, reported at a daily frequency.",
+    },
+    "film_number": {
+        "type": "string",
+        "description": "Document control number used in the SEC database.",
```
From what I found on Google: "The first four digits can be used to access scans of the document in the SEC's Virtual Private Reference Room." 

I'm not sure what else to include here. We could just drop this column if it doesn't seem useful.

zaneselvans replied:
If that's what the google dredged up that seems like good enough context. Maybe it's on microfische? Or was at one point.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "fiscal_year_end",
+                "sec10k_version",
+                "company_name_former",
+                "company_id_irs",
```
I'm not sure, I don't have any other information on this. I guess it's probably an EIN?

zaneselvans replied:
Looking up our old friend [Xcel Energy](https://eintaxid.com/company/410448030-xcel-energy-inc/) who has a TIN of `41-0448030`, it seems that that's what this column is, so I'd suggest we rename it to `taxpayer_id_irs` or `employer_id_irs`.

```python
(
    info_sec10k[info_sec10k.company_name.str.contains("xcel energy")
    .fillna(False)][["central_index_key", "company_name", "company_id_irs"]].head()
)
```
|   central_index_key | company_name    |   company_id_irs |
|--------------------:|:----------------|-----------------:|
|          0000072903 | xcel energy inc |        410448030 |
|          0000072903 | xcel energy inc |        410448030 |
|          0000072903 | xcel energy inc |        410448030 |
|          0000072903 | xcel energy inc |        410448030 |
|          0000072903 | xcel energy inc |        410448030 |

The IRS generally requires the 2-digit prefix (which includes leading zeroes) and the rest of the TIN to be separated by a dash, which would also keep these from accidentally getting cast to integers, and make it clearer that these are TINs.

It looks like there are 303 rows out of 250K non-null values that have fewer than the required 9 digits, 290 of which have 8 digits, so I wonder if they accidentally got converted to integers by Excel somewhere and lost a leading zero?  Interestingly there's only 1 of the 303 which has a leading zero.

Looking up [Duke Energy Services LLC](https://eintaxid.com/company/760632293-dcp-midstream%2C-llc/) however, it looks like the bad TIN is just a typo.

So maybe just re-format the 9-digit ones to XX-XXXXXXX format and leave the other ones alone.  Or we could set the non-conforming values to NA -- there are already 13,580 other NA values in the column.

katie-lamb replied:
I re-formatted the 9 digit ones to XX-XXXXXXX format and set the non-conforming ones to NA. Also added a regex constraint on the column and renamed to `taxpayer_id_irs`.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "company_name_former",
+                "company_id_irs",
+                "organization_name",
+                "sec_act",
```
Standardized with `df["sec_act"] = df["sec_act"].where(df["sec_act"].isnull(), "1934 act")` in the core table

katie-lamb commented on src/pudl/metadata/fields.py:
```diff
@@ -1757,6 +1761,10 @@
         "type": "string",
         "description": "EIA short code indicating the type of firing used by this boiler.",
     },
+    "fiscal_year_end": {
+        "type": "string",
+        "description": "The end date of an SEC filing company's fiscal year, in MMDD format.",
+    },
```
Added a constraint.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "sec10k_version",
+                "company_name_former",
+                "company_id_irs",
+                "organization_name",
```
I think it's probably not worth keeping this column.

katie-lamb commented on src/pudl/metadata/fields.py:
```diff
@@ -3416,6 +3424,10 @@
         "description": "Production expenses: water for power (USD).",
         "unit": "USD",
     },
+    "organization_name": {
+        "type": "string",
+        "description": "SEC standardized type of organization of the company.",
+    },
```
Decided in a different comment that it's probably not worth keeping this column.

katie-lamb commented on src/pudl/metadata/fields.py:
```diff
@@ -4197,6 +4209,14 @@
+        "type": "string",
+        "description": "SEC Act through which the form was enacted, e.g. 1934 act.",
+    },
+    "sec_file_number": {
+        "type": "string",
+        "description": "Filing number used internally by the SEC commission to track filing.",
+    },
```
Updated to `filing_number_sec`

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -39,13 +43,121 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+            "filing_values",
+            "former_company",
+        ],
+    )
```
Yeah I agree. I added `raise ValueError` if the values in the company information block column aren't equal to these listed values, but could log an error or do something else if there's a better way.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "central_index_key",
+                "report_date",
+                "company_name",
+                "name_change_date",
+                "company_name_former",
```
> when you look at the data, is the current name in row N the same as the former name in row N+1 for each company?

Ideally it would be, but often times a company reports a former name without there actually being a record extracted under that former name. So  about 30,000/50,000 CIKs are not duplicated in this change log.

We could fix this by making records for each of the former names where one isn't reported (like vertically stacking the former name column) and giving it a null `name_change_date`.

katie-lamb replied:
I went ahead and created records for the former names, concatenated with the existing, and dropped duplicates. There's still some weirdness, like this, where a piece of the time series of changes is never reported so the log doesn't make complete sense:
<img width="750" alt="Screenshot 2025-03-06 at 12 58 58 PM" src="https://github.com/user-attachments/assets/d347ad9f-a05f-4a74-b10e-7d920a11686d" />

But I don't think it's worth spending the time / is possible to reasonably reconcile that.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "phone_number",
+                "city",
+                "company_name",
+                "name_change_date",
```
Is there a way to make an intermediate `core` asset that's not persisted to the DB which has `name_change_date` and `former_name` so that it can be fed into the change log asset. But then the actual `core` table that's written to the DB doesn't have those columns?

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+    "out_sec10k__quarterly_company_information": {
+        "description": (
+            """Company information extracted from SEC10k filings and matched
+to EIA utilities using probabilistic record linkage. This table provides
+attributes about SEC 10k filing companies across time. The match between ``central_index_key``
+and ``utility_id_eia`` is one to one and doesn't change over time - the highest probability
+EIA utility match for each SEC company is used for all dates of reported information for
+that CIK."""
```
This PR doesn't include an association table between `central_index_key` and `utility_id_eia` but I can add one. This company information output table doesn't include any subsidiary companies, only companies whose information we extract from header blocks in filings, so I felt like speaking about subsidiary-only companies (without a CIK) here would be confusing. I still feel like this whole sentence is true:

> The match between ``central_index_key`` and ``utility_id_eia`` is one to one and doesn't change over time - the highest probability EIA utility match for each SEC company is used for all dates of reported information for
that CIK.

It could be made more specific by saying: 

> The match between ``central_index_key`` and ``utility_id_eia`` is one to one and doesn't change over time - the highest probability EIA utility match for each SEC company is used for all dates of reported information for that CIK. Subsidiary companies which don't file a SEC 10k are also linked to EIA utilities but are not reported in this table.

Thoughts?

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+                "phone_number",
+                "city",
+                "company_name",
+                "name_change_date",
```
See question about feeding an intermediate, not-persisted `core` table into the changelog and then having the `core` table not include these name change date and former name columns.

katie-lamb commented on src/pudl/metadata/resources/sec10k.py:
```diff
@@ -39,7 +39,105 @@
+        "description": (
+            """Company information extracted from SEC10k filings.
+This table provides attributes about SEC 10k filing companies across time.
+It represents a pivoted version of the raw company information table with extracted
+field values from the raw table as columns in this core table."""
```
Added more detail about the transformation in the docstring.

katie-lamb commented:
@zaneselvans Ok I think I've addressed all of the comments. Here are the remaining questions I have / unresolved issues:

* **Writing the `core_sec10k__quarterly_company_information` fails because of the following unique constraint:**

I changed the way we aggregate blocks of company information when transforming the raw data into the core table and now there are duplicate filenames in the core table. I get the following error:

```
dagster._core.errors.DagsterExecutionHandleOutputError: Error occurred while handling output "result" of step "core_sec10k__quarterly_company_information"::

sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: core_sec10k__quarterly_company_information.filename_sec10k
```

I can materialize the table when i use the default `io_manager_key`, but I'm not sure what to do about this error. `filename_sec10k` isn't part of the primary key for the table so I'm confused why it needs to be unique in the DB. 

* **Add an association table between `central_index_key` and `utility_id_eia` which doesn't include the associations between `subsidiary_id_sec10k` (or whatever that ID column will be called) and `utility_id_eia`. That subsidiary-only association table will go in with the next PR**

Also, see [this question](https://github.com/catalyst-cooperative/pudl/pull/4079#discussion_r1978700216) about how to rephrase the output table docstring to be more clear.

* **Remove `name_change_date` and `company_name_former` from the core table**
I currently create the change log asset by feeding in the core table and then taking the appropriate column. See [this question](https://github.com/catalyst-cooperative/pudl/pull/4079#discussion_r1984026899) about how to still have an upstream asset that feeds into the change log asset without needing keep the name change date and former name in the `core` table that's persisted to the DB.

* **Some weirdness in the change log table**
I updated the way that the change log table is created so that it better reflects an actual log of name changes. See [this comment](https://github.com/catalyst-cooperative/pudl/pull/4079#discussion_r1984052881) about some weirdness in reporting of names changes.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -48,7 +51,179 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+            "company_information_fact_name"
+        ].str.lstrip("]")
+    )
+    df = raw_sec10k__quarterly_company_information.pivot(
+        values="company_information_fact_value",
+        index=[
+            "filename_sec10k",
+            "report_date",
+            "company_information_block",
+            "company_information_block_count",
+        ],
+        columns="company_information_fact_name",
```
I added `company_information_block` and `company_information_block_count` to the index, then I groupby filename, report date, and company information block count to consolidate the information into one record per block.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -48,7 +51,179 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+        .drop(columns=["company_information_block", "company_information_block_count"])
+        .dropna(subset="central_index_key")
+    )
+    # we want central_index_key and report_date to be a primary key
+    # prioritize records where the filer is the same
+    # as the harvested central index key value
+    df["filer_cik"] = df["filename_sec10k"].str.split("/").str[2].str.zfill(10)
+    df["filer_cik_matches_cik"] = df["filer_cik"] == df["central_index_key"]
+    df = df.sort_values(by="filer_cik_matches_cik", ascending=False).drop_duplicates(
+        subset=["central_index_key", "report_date"], keep="first"
+    )
```
Interestingly, if you don't do this sorting before dropping duplicates on CIK and report dates then there are thousands of duplicate filenames + report dates. But when you do this sorting first there are only 2 duplicate filename + report dates. This felt wrong to me, but I couldn't figure out why it would be wrong. I think it's actually good - it means that almost all company information can be harvested directly from a filing submitted by the company itself.

jdangerx commented:
I get the following error:
> 
> ```
> dagster._core.errors.DagsterExecutionHandleOutputError: Error occurred while handling output "result" of step "core_sec10k__quarterly_company_information"::
> 
> sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: core_sec10k__quarterly_company_information.filename_sec10k
> ```
> 
> I can materialize the table when i use the default `io_manager_key`, but I'm not sure what to do about this error. `filename_sec10k` isn't part of the primary key for the table so I'm confused why it needs to be unique in the DB.

It looks like it was once the primary key for the table - the migration `migrations/versions/1e2ec7bf2b64_rename_sec_10k_tables_to_reflect_.py` has this on line 29:

```python
    sa.PrimaryKeyConstraint('filename_sec10k', 'filer_count', 'company_information_block', 'company_information_block_count', 'company_information_fact_name', 'company_information_fact_value', name=op.f('pk_core_sec10k__quarterly_company_information'))
    )
```

And then later in `migrations/versions/7901d29bb977_restructure_sec_company_info_tables.py` we drop `filer_count` and `company_information_block`:

```python
    with op.batch_alter_table('core_sec10k__quarterly_company_information', schema=None) as batch_op:
        < lots cut out...>
        batch_op.drop_column('company_information_fact_value')
        batch_op.drop_column('company_information_block')
        batch_op.drop_column('company_information_fact_name')
        batch_op.drop_column('company_information_block_count')
        batch_op.drop_column('filer_count')
```

But nowhere is the primary key constraint dropped.

This leads to the following schema in the database:

```
sqlite> .schema core_sec10k__quarterly_company_information
CREATE TABLE IF NOT EXISTS "core_sec10k__quarterly_company_information" (
        filename_sec10k TEXT,
        report_date DATE NOT NULL,
        central_index_key TEXT NOT NULL,
        phone_number TEXT,
        city TEXT,
        company_name TEXT,
        name_change_date DATE,
        film_number TEXT,
        fiscal_year_end TEXT,
        sec10k_version TEXT,
        company_name_former TEXT,
        sec_act TEXT,
        filing_number_sec TEXT,
        industry_id_sic TEXT,
        industry_name_sic TEXT,
        state TEXT,
        state_of_incorporation TEXT,
        street_address TEXT,
        address_2 TEXT,
        zip_code TEXT,
        zip_code_4 TEXT,
        taxpayer_id_irs TEXT,
        CONSTRAINT pk_core_sec10k__quarterly_company_information PRIMARY KEY (filename_sec10k)
);
```
Which shows that `filename_sec10k` is the only remaining primary key.

I think you'll need to either:
* downgrade past `1e2ec`, edit the migration to not include the primary key, and re-upgrade your database
* add a new migration (`alembic revision --autogenerate -m "remove primary key"` will generate an empty one for you) where you drop the constraints manually:
  ```python
def upgrade():
    with op.batch_alter_table('core...') as batch_op:
        batch_op.drop_constraint(...)

def downgrade():
     ... create constraint instead
```

katie-lamb commented:
@jdangerx thanks for looking into this! I went the route of creating a new migration and removing the primary key constraint and that seemed to have fixed it.

katie-lamb commented on src/pudl/analysis/pudl_models.py:
```diff
@@ -48,7 +51,193 @@ def core_sec10k__quarterly_company_information() -> pd.DataFrame:
+    return out_df
+
+
+@asset(io_manager_key="pudl_io_manager", group_name="sec10k")
+def core_sec10k__assn__sec10k_filers_and_eia_utilities() -> pd.DataFrame:
+    """An association table between SEC 10k filing companies and EIA utilities."""
+    matched_df = _load_table_from_gcs("out_sec10k__parents_and_subsidiaries")
+    matched_df = (
+        matched_df[["central_index_key", "utility_id_eia"]]
+        .dropna()
+        .drop_duplicates(
+            subset="central_index_key"
+        )  # matches should already be 1-to-1 but drop duplicates to ensure this is true
+    )
+    return matched_df
```
Instead of duplicating this code from the `out_sec10k__quarterly_company_information` table should I make the `out_sec10k__quarterly_company_information` asset a multi asset which creates two outputs - the association table and the output table?

jdangerx commented:
(accidentally posted this on #4097 instead)

Looks like there's only one outstanding major question here, but that maybe you two have stalled out a bit comms-wise. Let me see if, by adding a third party, we can get some clarity 😅 

Also, @zaneselvans IDK where you ended up with this after doing some work on it over the weekend.

Anyways. If the following sounds correct I think the path to getting this merged is:
* make an `_core...` asset that has the name change info, but isn't persisted
* make the existing `core_` assets depend on that
* rip out the `name_change_date`/`former_name` columns from all the persisted non-changelog tables

### `name_change_date`/`former_name` normalization issue
  
I'm inclined to agree with @zaneselvans that they should only live in the nice changelog table that @katie-lamb made.

I *think* @katie-lamb you're in agreement here that it would be neater to put the name change info only in the changelog - is this true?

I think if we still want to provide some notion of "hey, this company changed names at some point, beware" in the `out_` table we can use `last_name_change_date` and `most_recent_former_name` which then has a nice neat mapping to each (CIK, report_date) pair which we're using as the PK for each row. 

#### implementation question

It seems like @katie-lamb has a question about *how to implement* this - because the changelog depends on the `core_sec10k__quarterly_company_information` which is one of those (CIK, report_date) tables that we want to purge `name_change_date`/ `former_name` from.

One can make a `_core` asset that doesn't use the PUDL IO manager, then have both other `core_` tables depend on that `_core` asset.

While doing this, we may want the `_core` table columns to be the `last_name_change` and `most_recent_former_name` also - I'm not quite sure how you all are extracting this name change data and I'm running into a bit of trouble materializing locally.

zaneselvans reviewed (commented):
Interim comments that I failed to submit before getting bogged down in making actual changes...

zaneselvans commented:
Superseded by #4134
